### ISSUE TITLE
Model indexing slow working in big databases

### ISSUE THREAD

**Issue by rosenvladimirov** [odoo/odoo#144536]
Impacted versions: 
all version

Steps to reproduce:
create new model

Current behavior: 
in def check_indexes(self, cr, model_names):
1. do not create index for id field, this case the system stay very slow when database is grow up and more 1000000 records. When delete rows truncate cascade the operation cost to long time to search and found linked tables. Solution is indexes by id.
2. At moment all indexes is btree Access method, but for dynamic data with integer column in big tables cost lot of space and lot of time for searching inside.

Expected behavior:
1. To add rule: for id field to creating every time indexes in database.
2. To change Access method for integer field (like id) with brin.

little out of scope:
At 16+ version many read operation is moved in background, but this is ok for small databases, better solution is to use postgresql clusters, this up read speed minimum 4 times and profit from caching is very small. I think is better to invest time to create engine for clustering inside odoo machine for write operation. At moment do not have model to start in parallel on different machines writing process.

Video/Screenshot link (optional):
https://www.crunchydata.com/blog/postgres-indexing-when-does-brin-win

Support ticket number submitted via odoo.com/help (optional):

**Comment by pivi-odoo** [comment]
Hello @rosenvladimirov,

Thanks for your submittion.

The `id` column of all models is the Primary Key (Pkey), therefor by definition it is `unique`.
Postgresql enforces constrains of uniqueness via the creation of a `unique btree` index on the column. Therefor it is unnecessary to create an additional index of any kind on the `id` column alone.
As for using a BRIN index, those are really specific, since they are effective only when there is strong correlation between the distribution of the tuples and their on-disk page layout.
They are mostly useful for INSERT only models for columns that depends on the date of insertion (for example `create_date`).
In general you are better off just using a regular btree index, and only if the need if deemed necessary, then using a BRIN index for highly correlated fields would make sense.

**Comment by rosenvladimirov** [comment]
Hi @pivi-odoo,
All one2many connection use the inverse field and this field is not indexed. The brin indexes i used for "static" data fields, inverse field is always integer. The brin index is smaller and faster for big tables more 200k+.  On delete in one2many record from child side always search inverse fields. I test with brin and btree index for inverse fields, brin is more faster. Without index is very slow for big tables.

**Comment by pivi-odoo** [comment]
Hello @rosenvladimirov,

If you find a `Many2one` that is an inverse of a `One2many` in the standard code that isn't indexes (either `index=True` or `index='btree_not_null`), feel free to make a PR adding them, it's always possible that it's been forgotten.

As for the `BRIN` index, it is usually smaller yes, but be careful about evaluating it's performance, it's more effective *only* for data that is strongly correlated with their on-disk distribution. You can check the correlation with:
```sql
select correlation from pg_stats where tablename='your_table' and attname='your_column';
```
It should give you a number from `0` to `1`, if it's really close to `1`, then you have a strong correlation and a `BRIN` index will be better yes. But if it's not the case, it will be slower than a regular `btree`.

In a synthetic benchmark, usually since the table is created by generating rows from beginning to end, without any `UPDATE` / `DELETE`, you have a strong correlation, therefor a `BRIN` may perform better, but that might not be the case for a real-world use-case.

As of now Odoo doesn't support ORM side creation of all type of indexes Postgres has, but you are free to deploy them yourself on databases you find then effective on.

**Comment by rosenvladimirov** [comment]
Hello @pivi-odoo ,
i opened 2 PR: #173917 and  #174499, but i close today, there only discussions, i prepare monkey patch for my installations. For brin indexes, is my real experience with real data.
